examples/stdlib/stream/zip.effekt

```diff
@@ -0,0 +1,12 @@
+import stream
+
+def main() = {
+  def stream1() = [1, 2, 3].each
+  def stream2() = array::fromList(["a", "b", "c", "d"]).each
+
+  with val tup = for[(Int, String)] {
+    zip[Int, String] { stream1 } { stream2 }
+  }
+  val (a, b) = tup
```

Comment:
Perhaps `zip` should really exist as a multi-parameter `for`?
Then the code would be 
```suggestion
  with val (a, b) = for[Int, String] { stream1 } { stream2 }
```
... but then `for` has _a lot_ of different weird overloads (we already have a `for[A] {...} {...}` and `for[A, R] {...} {...}`, with this, we'd have also at least `for[A, B] {...} {...} {...}` and `for[A, B, R] {...} {...} {...}`)

---

Why not just `with val (a, b) = for ...`? See #707 to find out!

Reply:
I think it would be interesting to revisit @bracevac's work on event correlation for great motivating examples and inspiration: 

https://bracevac.org/assets/pdf/icfp18.pdf

Reply:
=> #716.